### ISSUE TITLE
Bad error message when trying to assign to member of reference type returned from function with deduced return type

### ISSUE THREAD

**Issue by davidstone** [llvm/llvm-project#74154]
The following ill-formed translation unit

```cpp
struct s {
	int m;
};

auto get(s const & x) {
	return &x;
}

void f(s const & x) {
	auto const ptr = get(x);
	ptr->m = 3;
}
```

causes clang to give an error message of

```console
<source>:11:9: error: cannot assign to variable 'ptr' with const-qualified type 'const const s *'
   11 |         ptr->m = 3;
      |         ~~~~~~ ^
<source>:10:13: note: variable 'ptr' declared const here
   10 |         auto const ptr = get(x);
      |         ~~~~~~~~~~~^~~~~~~~~~~~
1 error generated.
Compiler returned: 1
```

See it live: https://godbolt.org/z/WMMKo7hYe

This has multiple issues:

1. The type is printed incorrectly: it should print `const s * const`.
2. It claims I'm assigning to the variable `ptr`, when I am assigning to a member of the structure pointed to by `ptr`. This causes...
3. It says `ptr` was declared at a location other than its declaration.

**Comment by davidstone** [comment]
Note that a reference instead of pointer return type eliminates the `const const` issue but have the last two issues.

**Comment by shafik** [comment]
Yeah the gcc diagnostic is way better: https://godbolt.org/z/8GG8W393j

**Comment by elhewaty** [comment]
@shafik I am working on this issue.
I reached that the code responsible for emitting this error is: https://github.com/llvm/llvm-project/blob/d871f40deb7a7e6286b6d6b4476f6d30e444075e/clang/lib/Sema/SemaExpr.cpp#L14311-L14324
Can you give more hints?
thanks
